### ISSUE TITLE
Figure out variable visibility at block end instructions

### ISSUE THREAD

**Issue by saelo** [googleprojectzero/fuzzilli#233]
Currently, input variables for Block End instructions are allowed to come from the scope that is just being closed (see [here](https://github.com/googleprojectzero/fuzzilli/blob/10d5db4700b2d02481b916302269025cee9a1f69/Sources/Fuzzilli/FuzzIL/Code.swift#L194) for the algorithm). This is used e.g. for CodeStrings:

```
v6 <- BeginCodeString
    ...
    v9 <- ...
EndCodeString v9
```

which lifts to

```javascript
let s = `
    ...
    const v9 = ...
    v9;
`;
```

However, this behaviour is incorrect in other situations, e.g. for switch-case, as it would allow something like

```javascript
switch (v4) {
    case v2:
        const v5 = ...;
        break;
    case v5:
        ...
}
```

It seems that only allowing variables from the outer scope as inputs to block instructions is probably the right way to go. In that case, `Code.check()` needs to do `// Block and scope management (1)` before checking the input variables, and the [InputMutator](https://github.com/googleprojectzero/fuzzilli/blob/10d5db4700b2d02481b916302269025cee9a1f69/Sources/Fuzzilli/Mutators/InputMutator.swift#L25) would need to fetch a random variable from the outer scope (and the ProgramBuilder would need to expose such a method) for block end operations. Finally, the `CodeString` operation would probably need to change as well, in the easiest case simply dropping the input to the `EndCodeString` operation.

**Comment by amarekano** [comment]
I'll add a testcase for this as part of #216. I'll need some time to debug this further and get back to you with some ideas.

**Comment by amarekano** [comment]
I'm struggling to figure out how the switch case example you've listed previously would be generated.  Whenever we begin a switch case, the ScopeAnalyzer would end the previous scope and create a new one. So we shouldn't have a situation where a variable generated in a previous case block would be used as an input to the next case since it wouldn't be in the list of visible variables. Perhaps I'm missing something here?

**Comment by amarekano** [comment]
Here's the [commit](https://github.com/googleprojectzero/fuzzilli/commit/b1d232f22d98c8dc76df776bf92a443cb51f5db1#diff-1d91d056c0c48a8920a241ab83c54456e53c4adb82a2c45e0f8b5b22338259a2) that added an input variable to the `EndCodeString` operation. I'm not entirely sure why this was introduced. @samo98  would you happen to recall why we needed an input variable to `EndCodeString`?

**Comment by saelo** [comment]
> we may also have this issue with try-catch-finally blocks

I think there it should be fine, since the BeginCatch and BeginFinally instructions don't have any inputs (just outputs).

> I'm struggling to figure out how the switch case example you've listed previously would be generated.

It should happen for example due to a [mutation by the InputMutator](https://github.com/googleprojectzero/fuzzilli/blob/10d5db4700b2d02481b916302269025cee9a1f69/Sources/Fuzzilli/Mutators/InputMutator.swift#L31) of an existing switch-case.

>  I'm not entirely sure why this was introduced.

This is so that the CodeString always ends with an expression which will be the "return value" when passed to `eval`:
```javascript
let s = `
    ...;
    42;
`;
let r = eval(s);
// r === 42
```

But yeah, I think there are roughly three options for how to deal with this:

1. Ensure that inputs to inner block instructions always come from the outer scope. We'll need a `ProgramBuilder.randVarFromOuterScope` or similar API and use that in the InputMutator. Then we'll need to remove the input to the `EndCodeString` operation (I think that's currently the only one that relies on this)
2. Leave the semantics as they are now and always allow inputs to inner block instructions to come from the ending scope. With that, all inputs to e.g. switch-case, if-elseif-else, or similar constructs would have to be inputs of the starting instruction, so a switch-case would look like this `BeginSwitch v0, v1, v2, v3, v4; ... BeginSwitchCase; ... BeginSwitchCase; ...` (i.e. all case variables are inputs to the `BeginSwitch`
3. Let each operation decide about the allowed visibility of its inputs.

I think 3. would not be great, since it sounds like that would add some unnecessary complexity to the IL. Maybe 1. is the best option, but I'm open for suggestions.

**Comment by amarekano** [comment]
I've added a [commit](https://github.com/googleprojectzero/fuzzilli/pull/216/commits/e4992c576d2deba35107780e95c3b5ed37d09fe8) to implement item 1. Could you take a look please and let me know if this is how you'd like it to function.

**Comment by amarekano** [comment]
This issue is now being addressed with PR #258
